### ISSUE TITLE
Importing from data library should create a sync table always

### ISSUE THREAD

**Issue by saleiva** [CartoDB/cartodb#4775]
No matter if the user has sync tables or not, this dataset should be synced every 15 days. This should be transparent on the connecting workflow, but the user should be able to stop syncing from the dataset view.

cc/ @andrewxhill @Kartones

**Comment by andrewxhill** [comment]
my new favorite ticket of the day :)

cc @chriswhong because he has asked for it at least 3 times :)

**Comment by chriswhong** [comment]
I am going to -1 this and say there needs to be a user interaction to choose at the time of import, just like any other import.  Defaulting to sync may be annoying to people who are not expecting it.  What about free accounts who can't use sync tables?

**Comment by saleiva** [comment]
It says in the description. It should be activated even for accounts with no sync tables.
The user could always go to the dataset page and stop syncing.

S.

-----
@saleiva

**Comment by andrewxhill** [comment]
yeahhhh maybe chris is right we might have run a bit too fast. we should consider adding a modal on import from dat lib that just says (*) Sync ( ) Do not sync.

**Comment by chriswhong** [comment]
+1 a simple toggle is all that we need.

**Comment by saleiva** [comment]
That's the modal we have, but then we need to explain free users that they just can only sync data library datasets which is tricky... not sure guys...

**Comment by chriswhong** [comment]
I think there's no reason to enable it for free accounts, treat it the same
way as any other data.

It's also worth noting that sync frequency should be inherited from the
source data, not set to 15 days.  It should be inherited, and the user
would have an option to modify it or turn it off completely.

**Comment by saleiva** [comment]
Nah, I think that when talking about the data library sync should be
enabled for everybody. That would empower the dynamism on CartoDB and is
much better for the users, why shouldnt we benefit them with this?

Also changing the frequency when you don't control the source or when you
don't have a clear idea of how its being updated is weird, right?

I don't see any issue on syncying by default and letting users disconnect
those if they want.

**Comment by andrewxhill** [comment]
right, so i don't think we need to explain to them. it is just a feature
that we include with data library as a bonus :)

On Wed, Jul 29, 2015 at 12:01 PM, Sergio Álvarez Leiva <
notifications@github.com> wrote:

> Nah, I think that when talking about the data library sync should be
> enabled for everybody. That would empower the dynamism on CartoDB and is
> much better for the users, why shouldnt we benefit them with this?
>
> Also changing the frequency when you don't control the source or when you
> don't have a clear idea of how its being updated is weird, right?
>
> I don't see any issue on syncying by default and letting users disconnect
> those if they want.
> On Jul 29, 2015 5:54 PM, "Chris Whong" <notifications@github.com> wrote:
>
> > I think there's no reason to enable it for free accounts, treat it the
> same
> > way as any other data.
> >
> > It's also worth noting that sync frequency should be inherited from the
> > source data, not set to 15 days. It should be inherited, and the user
> > would have an option to modify it or turn it off completely.
> >
> > On Wednesday, July 29, 2015, Sergio Álvarez Leiva <
> > notifications@github.com>
> > wrote:
> >
> > > That's the modal we have, but then we need to explain free users that
> > they
> > > just can only sync data library datasets which is tricky... not sure
> > guys...
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125995077
> >.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125997376>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125999442>.
>



-- 

andrewxhill <http://twitter.com/andrewxhill>
http://cartodb.com

**Comment by saleiva** [comment]
So why to show it then? ;D

**Comment by andrewxhill** [comment]
so like chris says, they can remove the sync option if they don't want it.
my concern with the proposed flow, is that not many users would know to (1)
exit the visualization; (2) navigate to the Datasets; (3) find the newly
created dataset by the right name; (4) change the sync option

On Wed, Jul 29, 2015 at 12:35 PM, Sergio Álvarez Leiva <
notifications@github.com> wrote:

> So why to show it then? ;D
>
> On Jul 29, 2015 6:04 PM, "Andrew W. Hill" <notifications@github.com>
> wrote:
>
> > right, so i don't think we need to explain to them. it is just a feature
> > that we include with data library as a bonus :)
> >
> > On Wed, Jul 29, 2015 at 12:01 PM, Sergio Álvarez Leiva <
> > notifications@github.com> wrote:
> >
> > > Nah, I think that when talking about the data library sync should be
> > > enabled for everybody. That would empower the dynamism on CartoDB and
> is
> > > much better for the users, why shouldnt we benefit them with this?
> > >
> > > Also changing the frequency when you don't control the source or when
> you
> > > don't have a clear idea of how its being updated is weird, right?
> > >
> > > I don't see any issue on syncying by default and letting users
> disconnect
> > > those if they want.
> > > On Jul 29, 2015 5:54 PM, "Chris Whong" <notifications@github.com>
> wrote:
> > >
> > > > I think there's no reason to enable it for free accounts, treat it
> the
> > > same
> > > > way as any other data.
> > > >
> > > > It's also worth noting that sync frequency should be inherited from
> the
> > > > source data, not set to 15 days. It should be inherited, and the user
> > > > would have an option to modify it or turn it off completely.
> > > >
> > > > On Wednesday, July 29, 2015, Sergio Álvarez Leiva <
> > > > notifications@github.com>
> > > > wrote:
> > > >
> > > > > That's the modal we have, but then we need to explain free users
> that
> > > > they
> > > > > just can only sync data library datasets which is tricky... not
> sure
> > > > guys...
> > > > >
> > > > > —
> > > > > Reply to this email directly or view it on GitHub
> > > > > <
> > https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125995077
> > > >.
> > > > >
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <
> https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125997376
> > >.
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-125999442
> >.
> > >
> >
> >
> >
> > --
> >
> > andrewxhill <http://twitter.com/andrewxhill>
> > http://cartodb.com
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-126000423>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CartoDB/cartodb/issues/4775#issuecomment-126009934>.
>



-- 

andrewxhill <http://twitter.com/andrewxhill>
http://cartodb.com

**Comment by saleiva** [comment]
they can change it from the map view, plus, if we don’t want to expose tech stuff to non tech users, I think we shouldn’t add this step.

S.

-----
@saleiva

**Comment by chriswhong** [comment]
Dialing things back a bit, I think it's already fairly confusing that I can search my datasets and common data appears.  Keeping the common data library in its own separate search UI would emphasize the fact that it's a separate repository of data, and users could be a lot more deliberate with their searches.  (When I search for a US zipcode table I imported into my account, I always see Australian zip codes from common data)

I think a modal that says "This dataset has been set to sync every 24 hours from its external source.  Would you like your copy of this data to do the same?" Yes/No/Edit Frequency will do the trick, and also remind the user that they are cloning the table, not just "connecting" to it.

**Comment by andrewxhill** [comment]
it actually has the added benefit of advertising the sync capabilities more, which are right now underutilized imo.

at the end of the day though, it is your choice saleiva!

**Comment by Kartones** [comment]
Careful with this and enabling "selective syncs", that might have technical implications making this not a trivial change... 

I won't dig into what needs to be changed and coded to allow the feature until next sprint, but please define clearly this kind of details because syncs are right now binary: either you have them, or you don't. 

And IIRC for the Data Library there was already at least planned to be able to update them if the linked data library source changed (automatically/without sync tables), I'll ask @juanignaciosl for the details next week.

**Comment by saleiva** [comment]
Yeah I'll handle with Kartones. I think we are getting out of scope here :)

**Comment by juanignaciosl** [comment]
I'll hold my tongue about the feature discussion itself (I'd just like to add this might be an upselling opportunity for free users discovering sync potential ;-) ), but, from a technical point of view, there's a "improve data library update" issue, #3832, but it's not actually related, since it's just "updating data library listing". Making "a sync instead of an import" might not be trivial, as @Kartones said, but shouldn't be that hard either. There _might_ be a problem if a synched data library entry is updated, this scenario is something we should keep in mind when we do this.

**Comment by saleiva** [comment]
So yeah @kartones,

After doing some analysis we go with the following (which I think adds a bit of complexity but is amazing).

When selecting a data library dataset we just import it to your account as a sync table (updated every week). No extra step required. On the dataset list we show the datalibrary icon + the sync table indicator, so the user can still identify the dataset as part of the data library. To stop sincying, the user will be able to stop it anytime from the detail page or clicking the sync icon (creating a new issue for that).

The extra ball is: We should hide from the 'Data library' those datasets which you have imported to your account and that remain in sync. As soon as you stop syncing them we should show the dataset in your dashboard, plus restore it on the Data library (We could duplicate the dataset and add the date to the end of the name at this moment). I'm trying to avoid duplicates when searching datasets so if you have a data library dataset already in your account it doesn't appear again as part of the data library, since the data is the same.

**Comment by Kartones** [comment]
Ok, a few things from a quick glance yesterday afternoon at this:
- 1 week sync is not too agressive? those initial 15 days were more adequate, even if we don't download data because there are no news, in most cases datasets will be the same and we're still enqueueing sync jobs, etc. Is just to save resources
- We have some HUGE datasets, one is 850MB, a few ones are 250MB. Are we going to also sync those? 
- We have a platform limit set on maximum upload size, which by default is 150MB. This means you **cannot** sync any of the previous points dataset. And regarding that point I am not going to place a hack (an if) that skips this platform limit (except @javisantana mandates me to), because that's there for a reason and if we want syncs we play by the rules. So either we up the limits or only import small enough datasets or another solution...

**Comment by Kartones** [comment]
Branch: `4775-link-syncs-to-external-sources`

**Comment by Kartones** [comment]
curl to call a "data library sync" meanwhile UI fully supports it:
```bash
curl -i -H 'Content-Type: application/json' -H 'Accept: application/json' -X POST -d '{"type_guessing":"true","content_guessing":"true", create_vis":"false", "interval":"3600", "remote_visualization_id":"fc3acba4-400e-11e5-80a4-0800275dcf8f","value":"Continents"}' "http://kartones.localhost.lan:3000/api/v1/synchronizations?api_key=61ebf6cd0657862f2e3fd8d9b8654ed998a8f4e6"
```

**Comment by Kartones** [comment]
Temporally blocked due to the need to implement https://github.com/CartoDB/cartodb/issues/4996 to continue

**Comment by Kartones** [comment]
One thing, as agreed with @javisantana for now there will be the minimal backend checks regarding allowing or not this kind of syncs. Example in plain text: Code won't do specific checks about if you're trying to do a "sync now" of a remote-table-sync.

This is to not mess up normal code with too many special cases for remotes

**Comment by Kartones** [comment]
Note to self: Now that I'm unblocked with #4996 the way to link and filter already synced remote tables is:
```
SELECT visualizations(type:remote) v
-> external_sources(visualization_id: v.id) es
-> external_data_imports(synchronization_id:nil & external_source_id: es.id)
```

**Comment by Kartones** [comment]
Backend part initially done, waiting for possible fixes coming from implementing the frontend part (#5028)
